Matthias check puneets cmake branch

MatthiasHeilManchester opened the pull request:
Here it comes. Main changes are in README.md.  A few queries are embedded; long list of separate (but mainly straightforward (I think!) queries will arrive via mail.

MatthiasHeilManchester committed:
all four versions (mpi/tpl on/off) now work on mh's laptop. build script now interprets empty string in the _FROM variables as 'don't use that library'

MatthiasHeilManchester committed:
conflict with Puneet's fix to mumps inclusion; I decree that we don't need mpi

puneetmatharu committed:
Only run navier_stokes.bretherton in self-tests for debugging. DO NOT USE.

puneetmatharu committed:
Fix(?) navier_stokes.bretherton and update workflows.

 - Updated max residuals in navier_stokes.bretherton to 500.
 - Use macOS-15 for the macOS runner
 - Patch dependency version dumping in macOS GitHub workflow.

MatthiasHeilManchester committed:
updated .gitignore for MH's current config options; likely to be superseded by Puneet's version anyway

MatthiasHeilManchester committed:
Merge branch 'feature-shared-cmake' of github.com:puneetmatharu/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
working through the pre-beta test tidy up steps

- Reenabled two-level parallelism test in mpi airy_cantilever (had been commented out in validate.sh)
- removed "hierher"s introduced during move the cmake and associated debugging
- disabled doc time in mumps exact preconditioner

MatthiasHeilManchester committed:
Automated commit of clang-format CI changes.

MatthiasHeilManchester committed:
deal with warnings in demo-drivers (serial); make sure mumps is always built

MatthiasHeilManchester committed:
Merge branch 'matthias_check_puneets_cmake_branch' of github.com:MatthiasHeilManchester/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
resolving clash with puneet in yaml

MatthiasHeilManchester committed:
Don't use mumps as default solver

-- By default, we're disabling Mumps because it not only requires mpi
   (which would be easy to check at build time) but also requires MPI to
   have been initialised in the driver code before it can be used; not
   good for all the serial self-tests. The user can still set
   -DOOMPH_ENABLE_MUMPS_AS_DEFAULT_LINEAR_SOLVER=ON to override this setting.

-- Rewrote set_nboundary(...) to fix odd problem with resizing
   Boundary_coordinate_exists array (now doing it manually rather than
   using STL).

-- minor tweaks to make oomph-convert.py even more python3 compatible.

MatthiasHeilManchester committed:
Merge branch 'feature-shared-cmake' of github.com:puneetmatharu/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
Automated commit of clang-format CI changes.

MatthiasHeilManchester committed:
Merge branch 'feature-shared-cmake' of github.com:puneetmatharu/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
Few tweaks to the build machinery

-- Added defaults to all options in build script. This causes a slight
   problem default build of all third-party libraries (sensible, I think)
   then tries to build superlu_dist when there's no MPI, causing the build
   to fail. Changed external_distributions/CMakeLists.txt file to unset
   superlu_dist build if there's no mpi. Warning is provided.

-- Added option --root-OOMPH_OOMPH_INSTALL_DIR to oomph_build.py but couldn't
   figure out how to use it.

-- fixed wrong path in doc/poisson/two_d_poisson/two_d_poisson.txt ;
   generally documentation seems to build fine.

MatthiasHeilManchester committed:
Merge branch 'matthias_check_puneets_cmake_branch' of github.com:MatthiasHeilManchester/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
merged with Puneet and added (and spelled out) defaults in oomph_build.py

MatthiasHeilManchester committed:
Automated commit of cmake-format CI changes.

MatthiasHeilManchester committed:
nearly through the updated README.md; still work in progress

MatthiasHeilManchester committed:
Merge branch 'matthias_check_puneets_cmake_branch' of github.com:MatthiasHeilManchester/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
backup

MatthiasHeilManchester committed:
README.md rewrite done; time to proof read

MatthiasHeilManchester committed:
README.md over to Puneet

MatthiasHeilManchester committed:
tweak

MatthiasHeilManchester committed:
removed MH's old config options

MatthiasHeilManchester committed:
removed another reference to mh's build options

puneetmatharu commented on CMakeLists.txt:
```diff
@@ -95,15 +95,14 @@ oomph_option(OOMPH_ENABLE_PARANOID "Enable the PARANOID flag in Debug" ${HAVE_DE
-oomph_option(OOMPH_ENABLE_MUMPS_AS_DEFAULT_LINEAR_SOLVER "Use MumpsSolver as the default linear solver (when available)" ${OOMPH_DEFAULT_TO_MUMPS_SOLVER})
+# Choose the default linear solver. By default, we'll disable
+# Mumps because it not only requires mpi (which would be easy to
+# check) but also requires MPI to have been initialised in the driver
```
This seems a little problematic. What happened to the idea of initialising MPI in the constructor if it wasn't already set up?

MatthiasHeilManchester replied:
That would force us to "expose" all (serial) demo driver codes to mpi which would seem very unnatural (to me!). I think it's much better to have everything work in serial. If people want to use mumps (in an mpi enabled build) the addition of the single line linear_solver_pt() = new MUMPS; is all it takes do consciously make that switch. (OK, and initialise MPI, but it's all done deliberately).

puneetmatharu replied:
That makes sense. It's a bit annoying that it doesn't just work out of the box without MPI...

puneetmatharu commented on external_distributions/CMakeLists.txt:
```diff
@@ -167,10 +167,10 @@ endif()
-      FATAL_ERROR
-        "Cannot build SuperLUDist unless MPI is enabled (with -DOOMPH_ENABLE_MPI=ON)!"
-    )
+      WARNING
```
Hmm. This doesn't seem like a good idea. If the user specifically requests SuperLUDist, we shouldn't continue with the build if we can't actually build it. The warning message will likely just get swept up in the build output. Not to mention, if they're using `oomph_build.py` without the verbose flag, they'll never know there was a problem. The build should fail so the user can explicitly disable the build of SuperLUDist (since they can't build it). What's the intuition here for changing it to a warning?

MatthiasHeilManchester replied:
It's the other way around: If we want oomph_build.py to "just do it" (i.e. build everything) then we have to make it build superlu_dist too, unless there's no mpi. This is exactly what that switch achieves. 

The only problem would be if a user hand-configures oomh-lib and actively requests superlu_dist but doesn't do an mpi build. In that case they may be annoyed/confused that we've quiely dropped superlu_dist, but at least they'll end up with an otherwise working installation (and they can probably figure out that what they requested was a dumb combination; and the alert that we've change things is in the logs.

puneetmatharu replied:
Okay. The problem here is the updated `oomph_build.py` from your branch is not setting the default values correctly. It is not replicating the behaviour of the `CMakeLists.txt`. By default, it always tries to build SuperLUDist. However, it should be setting the default value to the value of `OOMPH_ENABLE_MPI` (which is what the `CMakeLists.txt` does. This isn't too difficult to do, but I also don't think it's the best way to proceed. I would suggest actually stripped out all instances of `default=...` in the Python script and only _describing the default value_ in the help string. That way when the user prints the help message, they know what the default is (which we still have to keep in line with the `CMakeLists.txt`), and the defaults (which make include dependent parameters) are left to the `CMakeLists.txt` to set.

MatthiasHeilManchester replied:
Let's discuss in person but I think I see what you're objecting to. cmake should indeed to what it's told; the change should therefore be made to the build script so that it doesn't request superlu_dist when mpi isn't requested.

MatthiasHeilManchester replied:
OK, all clear now (and I think you agreed to simply explain the default settings in oomph_build.py). Just give it a final check that my crimes have been undone.

puneetmatharu commented on oomph_build.py:
```diff
@@ -365,19 +365,20 @@ def expanded_path(p):
     # Flags common to both external_distributions and the root project
     common_group = parser.add_argument_group("common build flags")
-    common_group.add_argument("--OOMPH_ENABLE_MPI", metavar="ON/OFF", choices=["ON", "OFF"], help="Enable MPI in both external distributions and root project.")
+    common_group.add_argument("--OOMPH_ENABLE_MPI", default="OFF", metavar="ON/OFF", choices=["ON", "OFF"], help="Enable MPI in both external (third party) distributions and in oomph-lib; default: OFF")
```
Okay, I can definitely agree it's nice to provide defaults so the user knows what the default is, but now we need to make sure to always keep `oomph_build.py` consistent with the defaults in the `CMakeLists.txt` files. Are we actively agreeing to take on this responsibility?

MatthiasHeilManchester replied:
Responsibility accepted. By me!

puneetmatharu replied:
Then it's a deal!

puneetmatharu replied:
Having looked at the above problem with MPI and SuperLUDist, I think **we should not set the default values here**. Handling dependent values is fine but could quickly get messy. We should only _describe_ them in the help strings, e.g. 
```python
help="Build SuperLU_DIST in third-party libraries; default: same as --OOMPH_ENABLE_MPI (i.e. ON if MPI is ON, OFF if MPI is OFF)."
```
That way the defaults are set from the CMakeLists.txt file.

MatthiasHeilManchester replied:
All fine now; happy to close this. I've removed my (kept alive, but commented out) version from oomph_build.py. I think you agreed to explain all the default settings in the help string (and may have done so already).

puneetmatharu commented on src/generic/elastic_problems.cc:
```diff
@@ -569,27 +569,13 @@ namespace oomph
-#ifdef OOMPH_HAS_MPI
-    if (MPI_Helpers::mpi_has_been_initialised())
-    {
 #if defined(OOMPH_HAS_MUMPS) && \
```
I'm a little confused... I thought the problem was that MUMPS couldn't work without MPI being initialised? Can we do away with this check now? Apologies if I'm forgetting something we've already discussed!

MatthiasHeilManchester replied:
This is where it gets really complicated. There are a few statically built problems (e.g. this one, which allows the user to assign initial conditions for solid mechanics problems). It therefore gets built when the library is built, i.e. long before mpi is ever activated, so this is the only way to handle this. Again if the user doesn't like it, they could presumably overwrite the linear solver from the driver code but it's really quite an obscure little construction that's not worth agonising about.

puneetmatharu replied:
Okay, sounds messy and not worth it. I'm happy to accept these changes!

puneetmatharu commented on README.md:
````diff
@@ -69,75 +69,80 @@
-Make sure you have sufficiently recent versions of these programs installed on your computer:
+We assume you have checked out `oomph-lib` from its [GitHub repository](https://github.com/oomph-lib/oomph-lib). The document [`CONTRIBUTING.md`](CONTRIBUTING.md) provides detailed instructions for how to interact with GitHub, but to get started you can simply clone the repository onto your computer
+```bash
+git clone git@github.com:oomph-lib/oomph-lib.git
````
We should change this to using the simpler HTTPS over SSH. If you're just cloning the main repository, you won't be pushing to it directly, therefore you don't need to authenticate, thus you don't need the SSH method. If you've created a fork and need to push to your own fork, then it makes sense to use the SSH method, but it's better to reserve instructions on how to do that for the contributing guide

MatthiasHeilManchester replied:
??? This is what we say in CONTRIBUTING.md. Anyway, changed to

    git clone https://github.com/oomph-lib/oomph-lib.git

Hope that's what you meant.

puneetmatharu replied:
Yes, because when the user is contributing to oomph-lib, they have their own fork that they clone. When they push changes, it goes to their forked repo. But to do so, they need to push with authentication. Now assume a user clones the main `oomph-lib/oomph-lib` repository. They should never try to push changes, since it will be pushed directly to the main repository without a PR, which we do not allow. If the user has the sufficient powers, this could cause problems. To prevent this, it is better to use the HTTPS method so they are not automatically authenticated to push to the main repository. Does that make sense?

The change is also as desired, thanks!

puneetmatharu replied:
(I'll leave the thread unresolved for now so you can see my comment)

MatthiasHeilManchester replied:
All fine; I even understand it. (I think!). Pressing resolved myself.

puneetmatharu commented on README.md:
```diff
@@ -171,36 +178,122 @@ sudo apt install ninja-build
 
+### Compilers
+
+`oomph-lib` is written entirely in C++ and we require a compiler that can handle the C++17 standard. We also use third-party libraries written in C and Fortran so your compiler suite must be able to handle these too. On ubuntu you may have to install `gfortran`, using
```
Capitalise Ubuntu?

MatthiasHeilManchester replied:
Done

puneetmatharu commented on README.md:
```diff
@@ -171,36 +178,122 @@ sudo apt install ninja-build
+
+
 ## Building, installing and uninstalling `oomph-lib`
+`oomph-lib` relies on a number of third-party libraries. To facilitate the installation of these (as well as various optional third-party libraries), we provide the option to build them as part of our overall build process. We provide a detailed description of the two-stage build process below, but strongly encourage users to use our `oomph_build.py` script that performs all these actions in one  ...
```
Can we put tick marks around the `oomph_build.py` in the square brackets?

puneetmatharu replied:
Also should we replace "inexperienced" with "new"? Just to be a bit more *ahem* polite

puneetmatharu replied:
The link is also broken, I think you meant `#recommended-alternative-building-with-oomph_buildpy`

MatthiasHeilManchester replied:
Done (x3)

puneetmatharu commented on README.md:
```diff
@@ -171,36 +178,122 @@ sudo apt install ninja-build
+You'll see that by default we build the libraries in Release mode (i.e. with full optimisation) and in serial (i.e. without any MPI support), and then install them in `external_distributions/install`. This is done deliberately to make sure that even users who do not have root privileges on their computers can install the libraries.
+
+
+Given that these are third-party libraries it's unlikely that you'll want to build them in (slow) debug mode but you could do so by specifying the flag `-DCMAKE_BUILD_TYPE='Debug'` at the configure stage. If you have MPI installed on your machine and wish to use the parallel capabilities of some of the third-party libraries, specify `-DOOMPH_ENABLE_MPI='ON'`. You can also specify a different loca ...
```
Note that we don't need the `'` around the arguments, i.e. `-DCMAKE_BUILD_TYPE=Debug` is sufficient and so is `-DOOMPH_ENABLE_MPI=ON` (just for simplicity)

puneetmatharu replied:
Can we emphasise the importance of using an absolute path for the installation directory? (Which you've wisely used here.) In practice it is far better to use absolute paths to avoid problems with relative paths

MatthiasHeilManchester replied:
Done (x2 and everywhere)

puneetmatharu commented on README.md:
````diff
@@ -171,36 +178,122 @@ sudo apt install ninja-build
+```bash
+# Configure in debug mode and with MPI support. Use an existing OpenBLAS 
+# installation and install the others in a specific directory:
+cmake -G Ninja -B build \
````
Same thing here with being able to drop quotes unless the arguments have spaces in them

MatthiasHeilManchester replied:
done (everywhere)

puneetmatharu commented on README.md:
````diff
@@ -171,36 +178,122 @@ sudo apt install ninja-build
+# Now build/install it
+cmake --build build
+```
+**NOTE:** When specifying the directory that contains an existing third-party library, make sure that it contains the `lib` and `include` directories of the relevant library, so here
````
Maybe replace "so here" with "like here"?

MatthiasHeilManchester replied:
Not quite what I meant but rephrased.

puneetmatharu commented on README.md:
```diff
@@ -171,36 +178,122 @@ sudo apt install ninja-build
+total 12
+drwxrwxr-x 2 joe_cool  joe_cool  4096 May 27 11:53 include
+drwxrwxr-x 4 joe_cool  joe_cool  4096 May 27 11:53 lib
+
```
We can probably remove this empty trailing line in the code black for cleanliness

MatthiasHeilManchester replied:
Done (appreciate the obsession! :) )

puneetmatharu commented on README.md:
```diff
@@ -171,36 +178,122 @@ sudo apt install ninja-build
+
+
+### Step 2: Installing oomph-lib
+Now that the required third-party libraries are available, we proceed to the actual `oomph-lib` build process. Again we stress that inexperienced (or lazy) users may prefer to use the `oomph_build.py` script, described [below](#Building-with-oomph_build.py). For those who want to do the installation manually, we discuss three different options. The process is obviously very similar to the generic ...
```
Broken link again --> `#recommended-alternative-building-with-oomph_buildpy`

MatthiasHeilManchester replied:
Fixed (and searched for other instances; none found).

puneetmatharu commented on README.md:
````diff
@@ -171,36 +178,122 @@ sudo apt install ninja-build
 To configure, build and install the library, `cd` into the root directory of the cloned `oomph-lib` project (checked out from GitHub) and run the following commands:
 
 ```bash
 # Configure and generate the build system; -B specifies the
````
Since you've mentioned what the `-B` flag is for, should we mention that `-G` selects the choice of build system generator?

MatthiasHeilManchester replied:
Done

puneetmatharu commented on README.md:
```diff
@@ -171,36 +178,122 @@ sudo apt install ninja-build
+# build directory (here "build"). The final argument specifies
+# the location of the third-party libraries (built in the 
+# previous step).
+cmake -G Ninja -B build $(cat external_distributions/install/cmake_flags_for_oomph_lib.txt)
```
Should we add to the comment block that they should replace this with the path to their relevant installation directory? (if they chose a custom installation directory)

MatthiasHeilManchester replied:
Done!

puneetmatharu commented on README.md:
````diff
@@ -261,11 +353,62 @@ sudo ninja oomph_uninstall
+
+For instance, to build `oomph-lib` with MPI support, use
+```bash
+cmake -G Ninja -B build -DOOMPH_ENABLE_MPI='ON'
````
No `'` needed for simplicity

MatthiasHeilManchester replied:
Done (actually searched for all "=" so probably/hopefully found all offences of this type.

puneetmatharu commented on README.md:
````diff
@@ -261,11 +353,62 @@ sudo ninja oomph_uninstall
+```
+Note that by default we build the library in Release mode, so the above command will build a version of the library that is compiled with full optimisation and no debug support. 
+
+Other options are used mainly when developing new code. In this case, we recommend building with range checking (which makes the code run very slowly but detects array overruns) and "paranoia" which activates many internal self-tests throughout the library. Both will issue explicit error messages if problems are detected. However, these only tend to be useful if you can then reconstruct how the o ...
````
Hmm, if you're going to capitalise "Debug" in "Debug mode" here, you should do the same above (personally I agree with using the capitalised form)

puneetmatharu replied:
Should also wrap ddd in backticks, i.e. `ddd`

MatthiasHeilManchester replied:
Done (x2 and everywhere)

puneetmatharu commented on README.md:
````diff
@@ -261,11 +353,62 @@ sudo ninja oomph_uninstall
+
+We therefore recommend doing code development with the following configure options:
+```bash
+cmake -G Ninja -B build -DCMAKE_BUILD_TYPE='Debug' -DOOMPH_ENABLE_PARANOID='ON' -DOOMPH_ENABLE_PARANOID='ON'
````
Omit tick marks

MatthiasHeilManchester replied:
done

puneetmatharu commented on README.md:
````diff
@@ -261,11 +353,62 @@ sudo ninja oomph_uninstall
+```bash
+cmake -G Ninja -B build -DCMAKE_BUILD_TYPE='Debug' -DOOMPH_ENABLE_PARANOID='ON' -DOOMPH_ENABLE_PARANOID='ON'
+```
+The last two flags define the C++  macros `PARANOID` and `RANGE_CHECKING` which should be used as follows:
````
Double space between "C++" and "macros"

MatthiasHeilManchester replied:
done (or rather: gone)

puneetmatharu commented on README.md:
```diff
@@ -275,19 +418,19 @@ The `oomph_build.py` script is provided as a convenient one-stop solution for bu
 ### Using the script
 
-To use the build script, change into the top-level `oomph-lib` directory (the same directory that contains `oomph_build.py`) and run it with Python. For example:
+To use the build script, cd to the top-level `oomph-lib` directory (the same directory that contains `oomph_build.py`) and run it with Python. For example:
```
Need to wrap cd in backticks i.e. `cd`

MatthiasHeilManchester replied:
done

puneetmatharu commented on README.md:
````diff
@@ -275,19 +418,19 @@ The `oomph_build.py` script is provided as a convenient one-stop solution for bu
 ```
 
-By default, this will configure the project, compile all the libraries, and install them to the local `install/` directory (within the `oomph-lib` repository folder). The script will create a build directory (named `build/` by default) if one does not exist, or reuse it if it does. It combines the equivalent of the `cmake -B build`, `cmake --build build`, and `cmake --install build` commands into ...
+By default, this will build the third-party libraries, configure the `oomph-lib` project, compile its libraries, and install them to the local `install/` directory (within the `oomph-lib` root directory). The script will create a build directory (named `build/` by default) if one does not exist, or reuse it if it does. It thus combines the equivalent of the `cmake -B build`, `cmake --build build` ...
````
It will not reuse the build directory unless explicitly told to do so. If I added that line to the original text, my apologies. We throw an error if there is a build directory already and tell the user to explicitly reuse it if they want

MatthiasHeilManchester replied:
Yeah; should have spotted that! I think you wrote this before I suggested changing the behaviour. Clarified and made a forward reference to the command line options.

puneetmatharu commented on README.md:
```diff
@@ -275,19 +418,19 @@ The `oomph_build.py` script is provided as a convenient one-stop solution for bu
 After running `oomph_build.py`, if it completes without errors, `oomph-lib` should be built and installed in the specified location. You can then proceed to test the installation as described in the next section.
 
-> **Tip:** On Unix-like systems, you can make the script directly executable by running `chmod +x oomph_build.py`, then invoke it with `./oomph_build.py`. Using `python3 ...` or making it executable are equivalent as long as it's run with Python 3.
+> **NOTE:** On Unix-like systems, you can make the script directly executable by running `chmod +x oomph_build.py`, then invoke it with `./oomph_build.py`. Using `python3 ...` or making it executable are equivalent as long as it's run with Python 3.
```
Could change
> Using `python3 ...` or making it executable are equivalent as long as it's run with Python 3.

to 

> Using `python3 ...` is equivalent to using `./...` (after making the script executable).

If we're being extra picky

MatthiasHeilManchester replied:
done (and slightly rephrased).

puneetmatharu commented on README.md:
```diff
@@ -275,19 +418,19 @@ The `oomph_build.py` script is provided as a convenient one-stop solution for bu
-### Command-Line options
+### Command line options
 
 The `oomph_build.py` script supports several command-line options to customize its behavior. You can see a summary by running
```
"command-line" or "command line"? Need to be consistent between here and the heading

puneetmatharu replied:
"customise" with an "s"? For the British English spelling

MatthiasHeilManchester replied:
I think command-line is OK here because the three words are actually a compound noun. However, no excuses for the American spelling. Changed it.

puneetmatharu commented on README.md:
```diff
@@ -297,43 +440,45 @@ python3 oomph_build.py --help
-  > **Important:** When using this option, you will need to run the installation step with administrative privileges. The script will attempt to perform the installation step with sudo if possible, or it will remind you to re-run the script as root for the install phase. It’s generally recommended to run oomph_build.py `--root-CMAKE_INSTALL_PREFIX` under a normal user for the build, and let it pr ...
-- **`--wipe-tpl`**, **`--wipe-root`**, **`--wipe-doc`**: These options tell the script to remove the specified build/installation directories that would be written to when building the third-party libraries, root project, and documentation, respectively. Use the `--wipe-*` flags if you want a completely clean rebuild. For example
+- **`--root-OOMPH_ALLOW_INSTALL_AS_SUPERUSER`**: Use this option when you intend to install `oomph-lib` system-wide (by installing it into `/usr/local/`) using root privileges. This flag tells the script to configure the installation prefix to the system's default location (`/usr/local`) and to set any required internal CMake switches such as`OOMPH_ALLOW_INSTALL_AS_SUPERUSER=ON`. If you also spec ...
+  > **Important:** When using this option, you will need to run the installation step with administrative privileges. The script will attempt to perform the installation step with sudo if possible, or it will remind you to re-run the script as root for the install phase. It is generally recommended to run `oomph_build.py --root-CMAKE_INSTALL_PREFIX` as a normal user for the build, and let it prom ...
```
Backticks around sudo?

MatthiasHeilManchester replied:
Done; checked everywhere.

puneetmatharu commented on README.md:
```diff
@@ -297,43 +440,45 @@ python3 oomph_build.py --help
 
-  will delete the current `build/` directory and the `install/` directory (if they exist) before configuring a fresh build. `--wipe-doc` will remove the `doc/build/` directory.. These options are useful if you suspect a previous build is causing issues or if you want to reclaim space and rebuild from scratch. Warning: Wiping will permanently delete those directories (and all compiled files or ins ...
-- **`--reuse-tpl`**, **`--reuse-root`**, **`--reuse-doc`**: These flags are the counterpart to the `--wipe-*` options, explicitly instructing the script to reuse existing directories. Using `--reuse-*` flag means the script will allow the existing build/installation directories and will instead update or rebuild files as needed. These options can save time on iterative builds, but be mindful that ...
+  will delete the current `build/` directory and the `install/` directory (if they exist) before configuring a fresh build. `--wipe-doc` will remove the `doc/build/` directory. These options are useful if you suspect a previous build is causing issues or if you want to reclaim space and rebuild from scratch. Warning: Wiping will permanently delete those directories (and all compiled files or inst ...
```
`--wipe-doc` is a bit more involved; it not only wipes that build directory but it calls `git clean -xdf .` from the `doc/` directory. I.e. it uses Git for the clean-up process

EDIT: I'm now realising this requires more care because a bunch of `index.html` files also get generated in `demo_drivers` that are ignored by Git, so I didn't see that they needed to be cleaned up...

puneetmatharu replied:
Could probably make this a better warning by using `> [!WARNING]` followed by more lines starting with `>`, e.g.

> [!WARNING]
> Wiping will permanently delete those directories (and all compiled files or installed files therein), so use these flags with caution.

puneetmatharu replied:
Note that there are multiple variants of this including `[!TIP]`, `[!NOTE]`, `[!CAUTION]`, and `[!IMPORTANT]` which look like so:

> [!TIP]
> Blah blah blah

> [!NOTE]
> Blah blah blah

> [!CAUTION]
> Blah blah blah

> [!IMPORTANT]
> Blah blah blah

MatthiasHeilManchester replied:
Couldn't get this to work so I just added an old-skool **WARNING:**

puneetmatharu commented on README.md:
````diff
@@ -297,43 +440,45 @@ python3 oomph_build.py --help
+If the configuration is successful, you should see the message
 
 ```bash
 -- Build files have been written to: .../build
````
The triple dots might get confused for a relative path, should we change it with `[...]/build`?

MatthiasHeilManchester replied:
Done

puneetmatharu commented on README.md:
```diff
@@ -297,43 +440,45 @@ python3 oomph_build.py --help
 If there are errors (e.g., a required tool is not found or a configuration check failed), CMake will report them; the script will stop and propagate that error.
 
-Next, the script invokes the build stage (equivalent to running Ninja). If you specified the `--verbose` flag you will see compiler output as the `oomph-lib` libraries compile. This stage may take some time, especially on first build. If it succeeds, all necessary library files will be produced in the build directory.
+Next, the script invokes the build stage (equivalent to running Ninja). If you specified the `--verbose` flag you will see the compiler output as the `oomph-lib` libraries compile. This stage may take some time, especially on a first build. If it succeeds, all necessary library files will be produced in the build directory.
```
Hmm, we've "historically" said that the build process is invoked with `cmake --build build`, so we should say that instead of "Ninja" (which should actually be `ninja` -- lowercase -- since that refers to the command, as opposed to the build system).

MatthiasHeilManchester replied:
done.

puneetmatharu commented on README.md:
```diff
@@ -297,43 +440,45 @@ python3 oomph_build.py --help
-`oomph-lib` comes with an extensive list of well-documented example driver codes situated in the `demo_drivers/` directory. The driver codes in these folders are also used to validate the library.
+`oomph-lib` comes with an extensive list of well-documented example driver codes situated in the `demo_drivers/` directory. The driver codes in these directories are also used to validate the library.
 
 Once the library is installed you can (and should!) enter the `demo_drivers/` directory and run the self-tests as follows:
```
I can't comment in the hidden block below because it was part of a previous commit but we should change 

```bash
cmake -G Ninja -B build_for_demo_drivers
```

to

```bash
cmake -G Ninja -B build
```

for simplicity. We don't need to complicate it

MatthiasHeilManchester replied:
The command is actually preceded by an explicit comment:

 \# It's generally 
 \# a good idea to call it "build" but here we use
 \# a different name just to demonstrate that it has
 \# nothing to do with the build directory we used to
 \# build the library itself"

which was me talking to myself -- on behalf of other newbies who may also find the role of all these references to build (and the -B and --build and what not --- too many "b"s!) confusing. I'm voting to keep it in for that reason but if you object violently am happy to be overruled.

puneetmatharu replied:
Oops, then I blame myself for not reading it all! Let's leave it how you've done it.

puneetmatharu commented on README.md:
````diff
@@ -390,6 +535,17 @@ cd build
+cd demo_drivers/build
+ctest --rerun-failed --output-on-failure
+```
+To investigate/debug/fix the failed test, cd into the relevant demo_drivers directory and follow the instructions in the next section.
````
cd -> `cd`

puneetmatharu replied:
demo_drivers -> `demo_drivers`

MatthiasHeilManchester replied:
Done (x2)

puneetmatharu commented on README.md:
```diff
@@ -408,9 +569,62 @@ cd build
+Developers often wish to test if their latest changes to the library have broken specific driver codes. There are various mechanisms to identify potentially affected codes and to run the self-tests only on these:
+
+#### Filtering by TEST_NAME
+If you inspect the `CMakeLists.txt` files in the demo_drivers directories you'll notice that each executable is associated with a `TEST_NAME`, specified in the `oomph_add_test(...)` function. The test name usually mirrors the name of the directory containing the driver code (relative to the demo_drivers directory).  For example, in `demo_drivers/poisson/one_d_poisson/CMakeLists.txt` you will see  ...
```
demo_drivers -> `demo_drivers` (in both instances)

MatthiasHeilManchester replied:
Done (everywhere)

puneetmatharu commented on README.md:
```diff
@@ -408,9 +569,62 @@ cd build
+### Selective testing
+Developers often wish to test if their latest changes to the library have broken specific driver codes. There are various mechanisms to identify potentially affected codes and to run the self-tests only on these:
+
+#### Filtering by TEST_NAME
```
Hmm, should probably place TEST_NAME in backticks since it refers to code

MatthiasHeilManchester replied:
Done (everywhere)

puneetmatharu commented on README.md:
````diff
@@ -408,9 +569,62 @@ cd build
+  COMMAND ./validate.sh ${OOMPH_ROOT_DIR}
+  TEST_FILES validate.sh validata)
+```
+The TEST_NAME can be specified as a regular expression to `ctest`, using the `-R`/`--tests-regex` flag. Only tests for which the `TEST_NAME` key contains the specified regular expression will be run. CMake has its own conventions for forming regular expressions and we suggest you consult the relevant CMake documentation. However, for simple cases this is extremely straightforward; for example
````
First TEST_NAME isn't wrapped in backticks

MatthiasHeilManchester replied:
Done everywhere.

puneetmatharu commented on README.md:
````diff
@@ -408,9 +569,62 @@ cd build
+cd demo_drivers/build
+ctest -R 'poisson.one_d_poisson'
+```
+will run all the self tests whose TEST_NAME includes the string 'poisson.one_d_poisson'. This approach does, of course, assume that you know the relevant TEST_NAMEs. You could explore these for all the existing demo_driver codes by using the command
````
"self tests" -> "self-tests" to be consistent

puneetmatharu replied:
TEST_NAME -> `TEST_NAME`

puneetmatharu replied:
'poisson.one_d_poisson' -> `'poisson.one_d_poisson'`

puneetmatharu replied:
demo_driver -> `demo_drivers`

puneetmatharu replied:
I'll let you handle the remaining cases where backticks are required

MatthiasHeilManchester replied:
All done (I think!)

puneetmatharu commented on README.md:
````diff
@@ -408,9 +569,62 @@ cd build
+find . -name 'CMakeLists.txt' -exec grep -H TEST_NAME {} \;
+```
+#### Filtering tests by keywords appearing in demo driver codes
+Searching for relevant demo driver codes by their associated TEST_NAME, as explained in the previous section, is useful but the test name rarely includes all the features of the relevant code. A common scenario for developers is that, following changes to a specific class, they wish to test if all demo driver codes that explicitly use this class still work. To facilitate this we provide a helper  ...
````
TEST_NAME

MatthiasHeilManchester replied:
done

puneetmatharu commented on README.md:
````diff
@@ -408,9 +569,62 @@ cd build
+```bash
+scripts/filter_cmake_tests.py
+```
+that generates the `ctest` command required to run the self tests in all demo_drivers directories containing driver codes that contain a specified string. So, to generate the `ctest` command that tests all demo drivers that contain the string 'QAdvectionDiffusionElement', say, run the following command
````
demo_drivers -> `demo_drivers`

MatthiasHeilManchester replied:
done

puneetmatharu commented on README.md:
```diff
@@ -549,7 +763,25 @@ to your `.emacs` file will produce equivalent behaviour. You can now edit the so
 
-Developing your own code in an existing demo driver directory is a quick-and-dirty way to get started, especially since you are most likely to start your work by modifying an existing driver code. However, long-term this is not a sensible solution. One slightly more attractive alternative is to create a new directory, just for your code, in the `demo_drivers` directory; described further below. T ...
+
+Developing your own code in an existing demo driver directory is a quick-and-dirty way to get started, especially since you are most likely to begin by modifying an existing driver code anyway. However, long-term this is not a sensible solution. One slightly more attractive alternative is to create a new directory, just for your code, in the `demo_drivers` directory; described further [below](Cre ...
```
I think your link for "below" is broken; it doesn't contain a leading hash symbol

puneetmatharu replied:
"particulary" -> "particularly"

MatthiasHeilManchester replied:
done (x2)

puneetmatharu commented on README.md:
```diff
@@ -549,7 +763,25 @@ to your `.emacs` file will produce equivalent behaviour. You can now edit the so
+
+Developing your own code in an existing demo driver directory is a quick-and-dirty way to get started, especially since you are most likely to begin by modifying an existing driver code anyway. However, long-term this is not a sensible solution. One slightly more attractive alternative is to create a new directory, just for your code, in the `demo_drivers` directory; described further [below](Cre ...
+
+To illustrate the advantage of this approach, assume that you have a stand-alone driver code `non_oomph_lib_one_d_poisson.cc` that does not require the `oomph-lib` library. Using `g++` as the compiler you may want to compile this code using
```
A remark rather than a change: I really like this block. I think it's helpful to understand what build systems really do

puneetmatharu commented on README.md:
````diff
@@ -408,9 +569,62 @@ cd build
+# Disable
+set_tests_properties(poisson.one_d_poisson PROPERTIES DISABLED YES)
+```
+Disabled tests are listed explicitly in the final summary provided when `ctest` is run on the entire demo_drivers directory.
````
demo_drivers -> `demo_drivers`

MatthiasHeilManchester replied:
done

puneetmatharu commented on README.md:
```diff
@@ -610,18 +845,14 @@ oomph_add_executable(
+Note that this directory is completely unconnected to the `oomph-lib` directory. To be able to find the `oomph-lib` installation directory, if it is not in one of the system-wide standard locations such as `/usr/local`, it must be declared somehow.
 
-It is also possible to specify the location of the install directory in the `CMakeLists.txt` file, or to provide hints where to search for it. Please read the [CMake documentation of the `find_package()` function](https://cmake.org/cmake/help/latest/command/find_package.html?highlight=find_package).
+This is most easily done by using the `PATH` environment variable, as discussed [above](#option-2-specifying-a-custom-installation-location). Note that this is necessary even if `oomph-lib` is installed to its default installation directory, `install/`, in the `oomph-lib` root directory -- remember that the current directory is completely unconnected to the oomph-lib installation. Alternatively,  ...
```
Note: CMake doesn't check the `PATH` variable

puneetmatharu replied:
It only checks `PATH` when searching for a program, e.g.

```cmake
find_program(cmake ...)
```

MatthiasHeilManchester replied:
??? Too subtle for me. Does the text need to be changed?
Actually, I think your next comment clarified it. Please have a look to see if this is now correct.

puneetmatharu replied:
That's better, thanks! Although the "oomph-lib" in "the oomph-lib installation" needs backticks.

MatthiasHeilManchester replied:
Done.

puneetmatharu commented on README.md:
````diff
@@ -610,18 +845,14 @@ oomph_add_executable(
-Anyway, assuming `oomph-lib` was installed to `/home/joe_user/oomph_lib_install` the driver code can now be built using
+So, assuming `oomph-lib` was installed to `/home/joe_user/oomph_lib_install` the driver code can be built using
 
 ```bash
````
Option 1 won't work because `PATH` isn't read when locating libraries

MatthiasHeilManchester replied:
Ah, I see. Changed to:


\# Option 1: Define/export the oomphlib_ROOT variable and configure
export oomphlib_ROOT=/home/joe_user/oomph_lib_install
cmake -G Ninja -B build

which I think/hope is correct.

puneetmatharu replied:
Perfect!

puneetmatharu replied:
Note that you can specify `oomphlib_ROOT` as both a CMake flag and environment variable, e.g.

```bash
# Environment variables
# 1. Export and run
export oomphlib_ROOT=/home/joe_user/oomph_lib_install
cmake -G Ninja -B build
# 2. ...or feed directory into command
oomphlib_ROOT=/home/joe_user/oomph_lib_install cmake -G Ninja -B build

# CMake variable
cmake -G Ninja -B build -Doomphlib_ROOT=/home/joe_user/oomph_lib_install
```

See [here](https://cmake.org/cmake/help/latest/variable/PackageName_ROOT.html) and [here](https://cmake.org/cmake/help/latest/envvar/PackageName_ROOT.html#envvar:%3CPackageName%3E_ROOT).

MatthiasHeilManchester replied:
Actually I knew this (honest); not sure why my brain got muddled there. Thanks for spotting this.I think we can declare this to be resolved.

puneetmatharu commented on README.md:
```diff
@@ -652,11 +883,131 @@ mkdir RESLT
+
+**NOTE:** We provide a separate GitHub repository
+
+<div align="center">
```
Should we remove this? I think you mentioned merging this repository into the main oomph-lib repository in your email

MatthiasHeilManchester replied:
It's supposed to be a separate repository (just like the user's repo would be unrelated to oomph-lib). It's just hosted in the oomph-lib "space" on github "next to" oomph-lib itself.

puneetmatharu replied:
Ahhhhh! Okay, I had misunderstood. That's cleared it up. Thanks.

puneetmatharu commented on README.md:
```diff
@@ -652,11 +883,131 @@ mkdir RESLT
+### Customising driver codes
+
+#### Customising driver codes via the `oomph_add_executable()` function
+The `CMakeLists.txt` shows how to specify the sources and the name of the executable to be built via the arguments SOURCES and NAME, respectively. They provide the minimum information required by the compiler. You may, of course, wish to provide additional information to the compiler. For instance, if your code was a stand-alone code that does not not involve `oomph-lib` or any other third-party  ...
```
SOURCES and NAME could do with backticks

MatthiasHeilManchester replied:
Done (everywhere)

puneetmatharu commented on README.md:
````diff
@@ -652,11 +883,131 @@ mkdir RESLT
+```
+The aditional flags passed to the compiler are of two types:
+- `-Wall -Werror` are compiler flags that specify that the compiler should show all warnings and treat warnings as errors, respectively.
+- `-DREFINEABLE` passes the compiler macro REFINEABLE to the compiler (or, technically, to the preprocessor). Such macros are typically used to activate/de-activate parts of the C++ code, as in this code fragment
````
REFINEABLE -> `REFINEABLE`

MatthiasHeilManchester replied:
done

puneetmatharu commented on README.md:
```diff
@@ -652,11 +883,131 @@ mkdir RESLT
+  in the code.
+
+
+To accommodate such additional flags, the `oomph_add_exectutable()` function also accepts the optional flags
```
`exectutable` -> `executable`

MatthiasHeilManchester replied:
done

puneetmatharu commented on README.md:
```diff
@@ -652,11 +883,131 @@ mkdir RESLT
+
+To accommodate such additional flags, the `oomph_add_exectutable()` function also accepts the optional flags
+
+- `CXX_OPTIONS`: this should be used for compiler flags (e.g. `-Wall`, `-g`). Note that these tend to be compiler specific (i.e. not every compiler will use -g to compile with debug information). 
```
-g -> `-g`

MatthiasHeilManchester replied:
done

puneetmatharu commented on README.md:
````diff
@@ -652,11 +883,131 @@ mkdir RESLT
+  CXX_OPTIONS -Wall -Werror
+  CXX_DEFINITIONS REFINEABLE)
+```
+will replicate the g++ command shown above (but also link the code against `oomph-lib` and any associated third-party libraries). 
````
g++ -> `g++`

MatthiasHeilManchester replied:
done

puneetmatharu commented on README.md:
````diff
@@ -652,11 +883,131 @@ mkdir RESLT
+```
+will replicate the g++ command shown above (but also link the code against `oomph-lib` and any associated third-party libraries). 
+
+**NOTE:** Given that most compilers pass macros via the -D option, you could also add the flag `-DREFINEABLE` to the CXX_OPTIONS and omit the CXX_DEFINITIONS. It will achieve the same thing but may upset the CMake purists.
````
-D -> `-D`

puneetmatharu replied:
Backticks around the CXX_* variables

puneetmatharu replied:
> may upset the CMake purists.

...guilty as charged!

MatthiasHeilManchester replied:
done

puneetmatharu commented on README.md:
```diff
@@ -652,11 +883,131 @@ mkdir RESLT
+
+#### Customising targets using native CMake commands; hashed target names
+
+If you are comfortable with CMake, you may wish to control the target properties of executables in a `CMakeLists.txt` file using native CMake commands. When doing this it is important to realise that the NAME specified in the call to `oomph_add_executable(...)` is not the name used by CMake itself. (The reason is technical: we create a modified, hashed version of the name which includes the path  ...
```
`NAME`

puneetmatharu replied:
`demo_drivers`

MatthiasHeilManchester replied:
done (everywhere)

puneetmatharu commented on README.md:
````diff
@@ -652,11 +883,131 @@ mkdir RESLT
+target_compile_definitions(one_d_poisson PRIVATE ${OOMPH_COMPILE_DEFINITIONS})
+```
+
+*** hierher Puneet:  PRIVATE (or some equivalent) seems to be required; also this doesn't work ! ***
````
Ah this is an interesting one. Nice spot! The reason this will fail is later in `oomph_add_test`, when it looks for the hashed name of the executable. 

If you use 
```cmake
# Specify the target name and source(s)
add_executable(one_d_poisson one_d_poisson.cc)

# Link against oomph-lib's poisson library
target_link_libraries(one_d_poisson PRIVATE oomph::poisson oomph::meshes oomph::generic)

# Compiler options
target_compile_options(one_d_poisson PRIVATE -Wall -Werror)

# Preprocessor macros
target_compile_definitions(one_d_poisson PRIVATE REFINEABLE)

# Pull in the information about oomph-lib build
target_compile_definitions(one_d_poisson PRIVATE ${OOMPH_COMPILE_DEFINITIONS})
```

then you'll need to pass the `ASSUME_PURE_CMAKE_TARGET_NAMES` flag to `oomph_add_test` (note that I have only just this second added that flag; you'll need to pull those changes in from my branch)

```cmake
oomph_add_test(
  TEST_NAME poisson.one_d_poisson
  DEPENDS_ON one_d_poisson
  COMMAND ./validate.sh ${OOMPH_ROOT_DIR}
  TEST_FILES validate.sh validata
  ASSUME_PURE_CMAKE_TARGET_NAMES)
```

MatthiasHeilManchester replied:
let's discuss "in person" to make sure I understand.

MatthiasHeilManchester replied:
Resolved as far as I'm concerned.

puneetmatharu commented on README.md:
```diff
@@ -723,7 +1074,7 @@ cmake -G Ninja -B build -DCMAKE_MESSAGE_LOG_LEVEL=VERBOSE
 
```
It's worth mentioning that (for the line in the hidden block above) the default log level is verbose already

MatthiasHeilManchester replied:
Don't understand your comment (and missed it earlier).

puneetmatharu replied:
Ignore me. I'm confusing myself. The log-level is only set to verbose by default for the `external_distributions` and main project; you have to quite rightly specify it for the demo drivers. Sorry!

puneetmatharu commented on README.md:
````diff
@@ -1050,91 +1401,33 @@ ninja install
+To uninstall the documentation do
 ```bash
-cmake --list-presets
+ hierher Puneet -- help!
````
This is messier than I thought. It's worth us having a discussion about what exactly is generated and how we spot it. It's not, e.g. just ALL `index.html` files since `picture_show/index.html` (I believe) is a Git-tracked file that we should not delete, and doing `git clean -xdf .` in a directory other than `doc/` is dangerous (since the user might have files we don't want to delete)

MatthiasHeilManchester replied:
Sure. I'll keep this comment alive.

puneetmatharu commented on README.md:
```diff
@@ -1050,91 +1401,33 @@ ninja install
-We recommend that you can write your own `CMakeUserPresets.json` file. You can inherit your presets from the presets we provide in `CMakePresets.json`. For details on how to do this refer to the [CMake documentation](https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html).
-
-**Remark:** We recommend that you do not use the Ninja Multi-Config generator yet.
+To add new tutorials to the doc directory, follow the same steps used for adding new demo drivers.
```
doc -> `doc`

MatthiasHeilManchester replied:
done

puneetmatharu commented on README.md:
````diff
@@ -1050,91 +1401,33 @@ ninja install
-```
 
 ## Cheat sheet: autotools vs. cmake
+Earlier releases of `oomph-lib` used the autotools to build/install the library. To ease the conversion to the new way of doing things, the table below provides a side-by-side comparison of key operations. Note that we also provide a script, `scripts/***py`, ***hierher Puneet*** that (attempts to) converts an autotools `Makefile.am` into a corresponding `CMakeLists.txt`. It may need a bit work/ch ...
````
"used the autotools" -> "used Autotools"

puneetmatharu replied:
The conversion script is `scripts/oomph_generate_cmake_script.py` but it's a couple of years old and doesn't account for every possible edge case but it's probably a good starting point. It's possibly worth you testing it out on your own private code to see if it works

MatthiasHeilManchester replied:
OK. Done (x2).

puneetmatharu commented on README.md:
```diff
@@ -1050,91 +1401,33 @@ ninja install
 
-Assume that `$oomph_home_dir` is the `oomph-lib` home directory:
+In all the examples shown below we assume that `$oomph_home_dir` is the `oomph-lib` home directory:
 <table>
```
Hmm, this table is really hard to view in the browser; it's too wide to fit easily on the screen. Not to mention it's not easy for a user just reading the RAW markdown to navigate in their text editor.  If we keep each table entry as a heading then it can also be navigated to from the table of contents

MatthiasHeilManchester replied:
Looks OK-ish for me. Messed around with this quite a bit and really wanted the side-by-side layout. Not sure what else to do. I think if anybody's brave enough to read the raw markdown they'll realise soon enough to that this is not intended to be consumed in that form. So I propose we keep it in that form (unless you have any clever ideas how to fix this.

puneetmatharu replied:
My proposal was that each block in the table become a subheading then we have

```txt
### HOW TO DO XYZ

#### AUTOMAKE
...code...

#### CMake
...code...
```

Can be discussed on a call!

puneetmatharu commented on README.md:
````diff
@@ -1417,120 +1734,76 @@ cmake --install build<br>
-# Do something to the target...
-set_target_properties(${OOMPH_TARGET_NAME} PROPERTIES CXX_STANDARD 20)
-```
+---
````
So am I! Let's kill it

MatthiasHeilManchester replied:
Killed.

puneetmatharu reviewed (changes requested):
Phew! This is a huge amount of work (from your side) and it's looking amazing. I'm excited to get this out the door! Thank you for handling this. I've put a number of comments in but they're mostly minor, e.g. typos, adding backticks, etc.

MatthiasHeilManchester committed:
dealt with most of Puneet's change requests for README.md

MatthiasHeilManchester committed:
fpdiff tweaks from the office machine

MatthiasHeilManchester committed:
merged Puneet's changes; still need to sort out defaults in oomph_build.py; left hierhers

MatthiasHeilManchester committed:
Sorted out the warnings about vector<bool>.

Done mainly by replacing the

  std::vector<bool> Mesh::Boundary_coordinate_exists;

with

  Vector<unsigned> Mesh::Boundary_coordinate_exists_stored_as_unsigned;

with suitable access functions rather than direct access, so

  Boundary_coordinate_exists[b] = true;

becomes

  set_boundary_coordinate_exists(b);

Similarly

   Boundary_coordinate_exists[b]

becomes

   boundary_coordinate_exists(b)

etc.

MatthiasHeilManchester committed:
Automated commit of clang-format CI changes.

MatthiasHeilManchester committed:
Merge branch 'feature-shared-cmake' of https://github.com/puneetmatharu/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
added user_drivers/joe_cool

MatthiasHeilManchester committed:
Automated commit of cmake-format CI changes.

MatthiasHeilManchester committed:
explaine user_drivers; update .gitignore

MatthiasHeilManchester committed:
Merge branch 'matthias_check_puneets_cmake_branch' of github.com:MatthiasHeilManchester/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
cmake/OomphGenerateDocFrom.cmake tweaks; remote controlled by Puneet on zoom

MatthiasHeilManchester committed:
merged with puneet; had a clash in cmake/OomphGenerateDocFrom.cmake; used my version with backslashes at ends of lines

MatthiasHeilManchester committed:
more README.md; removed user_drivers from top-level CMakeLists.txt; drivers can't be built before library is installed. Doh!

MatthiasHeilManchester committed:
retain oomph-lib-config.h but remove reference to autoconfig

MatthiasHeilManchester committed:
Automated commit of clang-format CI changes.

MatthiasHeilManchester committed:
tweaks

MatthiasHeilManchester committed:
Merge branch 'matthias_check_puneets_cmake_branch' of github.com:MatthiasHeilManchester/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
Merge branch 'feature-shared-cmake' of github.com:puneetmatharu/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
merged with Puneet

MatthiasHeilManchester committed:
Puneet's changes to cmake/OomphGenerateDocFrom.cmake to fix the doc build problem

MatthiasHeilManchester committed:
merged

MatthiasHeilManchester committed:
Automated commit of cmake-format CI changes.

puneetmatharu commented on README.md:
```diff
@@ -1760,6 +1765,27 @@ cmake --install build<br>
+With these flags, the header files in the `install` directories (usually copied from the `src` directory) are replaced by symbolic links to the original files in the `src` directory. This means that, if you build a library and are alerted to an error in a header file you don't accidentally edit a copy (that is then duly overwritten when you reinstall the library).
+
+### How to update third-party libraries to later versions
+The third-party libraries in `external_distributions` are pulled in from their respective GitHub repositories. The version of the libraries is typically encoded in the `<library_name>_TARBALL_URL` variable name in the files `external_distributions/cmake/OomphGetExternal<library_name>.cmake` file. An upgrade should (usually) just require a change to the tarball name, though if a more recent versio ...
```
### How to update third-party libraries to later versions

The third-party libraries in `external_distributions` are pulled in from their respective GitHub repositories. The version of the libraries is typically encoded in the `<library_name>_TARBALL_URL` variable name in the files `external_distributions/cmake/OomphGetExternal<library_name>.cmake` file. An upgrade should (usually) just require a change to the tarball name or Git tag, though if a more recent version of the library requires different build steps the relevant `*.cmake` will have to be modified further. Most external packages live in `external_distributions` and are downloaded and built automatically. To switch to a newer release (or to add a brand-new package) you normally touch three places:

| Where                                                                           | What it does                                                                      | What you must change                                                                                     |
|---------------------------------------------------------------------------------|-----------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------|
| `OomphGetExternal\<Lib\>.cmake` (in `external_distributions/cmake/`)            | Downloads & builds the package with FetchContent/ExternalProject                  | Update the `TARBALL_URL` or `GIT_TAG` if a Git clone is used.                                            |
| `OomphLocateThirdPartyLibs.cmake` (in `external_distributions/cmake/`)          | Tells the main `oomph-lib` build where to find the freshly built library          | Search for the `find_package(<PACKAGE> ...)` call and update the version component.                      |
| `oomphlibConfig.cmake.in` (in `src/cmake/`); installed as`oomphlibConfig.cmake` | Exposes `oomph-lib`'s build to other projects via `find_package(oomphlib CONFIG)` | Change the relevant package version (in a `find_dependency(<PACKAGE> ...)` call) to what you used above. |

#### Minimal example – bumping CGAL 6.0.1 -> 6.2.0

**`OomphGetExternalCGAL.cmake`**:

```bash
# OLD
set(CGAL_TARBALL_URL https://github.com/CGAL/cgal/releases/download/v6.0.1/CGAL-6.0.1.tar.xz)

# NEW
set(CGAL_TARBALL_URL https://github.com/CGAL/cgal/releases/download/v6.2.0/CGAL-6.2.0.tar.xz)
```

**`OomphLocateThirdPartyLibs.cmake`**:

```cmake
# OLD
find_package(CGAL 6.0.1 REQUIRED GLOBAL PATHS ${OOMPH_USE_CGAL_FROM} NO_DEFAULT_PATH)

# NEW
find_package(CGAL 6.2.0 REQUIRED GLOBAL PATHS ${OOMPH_USE_CGAL_FROM} NO_DEFAULT_PATH)
```

**`oomphlibConfig.cmake.in`**:

```cmake
# OLD
find_dependency(CGAL 6.0.1 REQUIRED GLOBAL PATHS ${OOMPH_USE_CGAL_FROM} NO_DEFAULT_PATH)

# NEW
find_dependency(CGAL 6.2.0 REQUIRED GLOBAL PATHS ${OOMPH_USE_CGAL_FROM} NO_DEFAULT_PATH)
```

MatthiasHeilManchester committed:
hacking around doc; don't do diff against this version

MatthiasHeilManchester committed:
merged with puneet

MatthiasHeilManchester committed:
Merge branch 'matthias_check_puneets_cmake_branch' of github.com:MatthiasHeilManchester/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
messing with drop down menus

MatthiasHeilManchester committed:
Merge branch 'feature-shared-cmake' of github.com:puneetmatharu/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
merge with Puneet

MatthiasHeilManchester committed:
Automated commit of clang-format CI changes.

MatthiasHeilManchester committed:
tidy up doc (getting there...); some doxygen warnings may have to stay....

MatthiasHeilManchester committed:
Merge branch 'matthias_check_puneets_cmake_branch' of github.com:MatthiasHeilManchester/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
Automated commit of clang-format CI changes.

MatthiasHeilManchester committed:
Automated commit of cmake-format CI changes.

puneetmatharu commented on README.md:
````diff
@@ -413,6 +418,14 @@ double some_function(const double& x)
+
+Finally, developers should install the library headers as symbolic links to the actual header files. If you don't do this you'll go insane while fixing bugs in copies of header files that immediately get over-written when the (supposed to be fixed) library is reinstalled. So install the library as follows if you intend to work on it.
+```bash
+cmake -G Ninja -B build -E create_symlink`
````
This doesn’t really make sense unfortunately. This is mixing a configuration with the command mode (`-E`). You need to use `-DOOMPH_INSTALL_HEADERS_AS_SYMLINKS=ON`

puneetmatharu commented on README.md:
```diff
@@ -1781,8 +1797,65 @@ cmake -G Ninja -B build -E create_symlink`
+
+#### A self-test fails even though the output files produced by the code are correct
+Self-tests are performed by
+the \c validate.sh shell script, which runs the executable and
```
Just a heads up that there’s a bunch of Doxygen style markup (which I’m sure you probably already know about, but it’s easily forgotten…)

puneetmatharu commented on README.md:
```diff
@@ -1781,8 +1797,65 @@ cmake -G Ninja -B build -E create_symlink`
+
 ### How to update third-party libraries to later versions
-The third-party libraries in `external_distributions` are pulled in from their respective GitHub repositories. The version of the libraries is typically encoded in the `<library_name>_TARBALL_URL` variable name in the files `external_distributions/cmake/OomphGetExternal<library_name>.cmake` file. An upgrade should (usually) just require a change to the tarball name, though if a more recent versio ...
+The third-party libraries in `external_distributions` are pulled in from their respective GitHub repositories. The version of the libraries is typically encoded in the `<library_name>_TARBALL_URL` variable name in the files `external_distributions/cmake/OomphGetExternal<library_name>.cmake` file. An upgrade should (usually) just require a change to the tarball name, though if a more recent versio ...
```
Hmm, they’re not all pulled from their respective repos. Some are, but some I uploaded to the oomph-lib release so we can ship a stable tarball (incase their repo is mysteriously set alight at night), I just didn’t get around to doing this for all of them.

puneetmatharu reviewed (changes requested):
Just a few (possibly annoying or) possibly helpful comments!

MatthiasHeilManchester committed:
more doxygen debugging; picture show fixed

MatthiasHeilManchester committed:
Merge branch 'matthias_check_puneets_cmake_branch' of github.com:MatthiasHeilManchester/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
Automated commit of clang-format CI changes.

MatthiasHeilManchester committed:
more doc tweaks

MatthiasHeilManchester committed:
Merge branch 'matthias_check_puneets_cmake_branch' of github.com:MatthiasHeilManchester/oomph-lib into matthias_check_puneets_cmake_branch

MatthiasHeilManchester committed:
duplicated </a> in mesh list; but doesn't fix it

puneetmatharu committed:
Rewrite external_distributions and remove old sources:
 - Use Git repos (and tag versions) instead of tarballs.
 - Patch cmake/scripts/run_cgal_self_test.sh.
 - Remove space_time/space_time_navier_stokes/archived_discretisations.

puneetmatharu committed:
Automated commit of clang-format CI changes.

MatthiasHeilManchester committed:
merged with Puneet

MatthiasHeilManchester committed:
tweaks to doc; mesh list processed with chatgpt suggestions but still broken

MatthiasHeilManchester committed:
updated links to MH's old papers
